qubes/storage/callback.py

```diff
@@ -0,0 +1,442 @@
+        assert isinstance(conf_all, dict), 'The file %s is supposed to define a dict.' % CallbackPool.config_path
+
+        try:
+            self._cb_conf = conf_all[self._cb_conf_id]
+        except KeyError:
+            #we cannot throw KeyErrors as we'll otherwise generate incorrect error messages @qubes.app._get_pool()
+            raise NameError('The specified conf_id %s could not be found inside %s.' % (self._cb_conf_id, CallbackPool.config_path))
+
+        try:
+            bdriver = self._cb_conf['bdriver']
+        except KeyError:
+            raise NameError('Missing bdriver for the conf_id %s inside %s.' % (self._cb_conf_id, CallbackPool.config_path))
+
+        self._cb_cmd_arg = json.dumps(self._cb_conf, sort_keys=True, indent=2)
+
+        try:
+            cls = qubes.utils.get_entry_point_one(qubes.storage.STORAGE_ENTRY_POINT, bdriver)
+        except KeyError:
+            raise NameError('The driver %s was not found on your system.' % bdriver)
+        assert issubclass(cls, qubes.storage.Pool), 'The class %s must be a subclass of qubes.storage.Pool.' % cls
+
+        self._cb_requires_init = self._check_init()
+        bdriver_args = self._cb_conf.get('bdriver_args', {})
+        self._cb_impl = cls(name=name, **bdriver_args)
+
+        super().__init__(name=name, revisions_to_keep=int(bdriver_args.get('revisions_to_keep', 1)))
+        self._cb_ctor_done = True
+        self._callback('on_ctor')
+
+    def _check_init(self):
+        ''' Whether or not this object requires late storage initialization via callback. '''
+        cmd = self._cb_conf.get('on_sinit')
+        if not cmd:
+            cmd = self._cb_conf.get('cmd')
+        return bool(cmd and cmd != '-')
+
+    def _init(self, callback=True):
+        #late initialization on first use for e.g. decryption on first usage request
+        #maybe TODO: if this function is meant to be run in parallel (are Pool operations asynchronous?), a function lock is required!
+        if callback:
+            self._callback('on_sinit')
+        self._cb_requires_init = False
+
+    def _assertInitialized(self, **kwargs):
+        if self._cb_requires_init:
+            self._init(**kwargs)
+
+    def _callback(self, cb, cb_args=[], log=logging.getLogger('qubes.storage.callback')):
+        '''Run a callback.
+        :param cb: Callback identifier string.
+        :param cb_args: Optional arguments to pass to the command as first arguments.
+                        Only passed on for the generic command specified as `cmd`, not for `on_xyz` callbacks.
+        '''
+        if self._cb_ctor_done:
+            cmd = self._cb_conf.get(cb)
+            args = [] #on_xyz callbacks should never receive arguments
+            if not cmd:
+                cmd = self._cb_conf.get('cmd')
+                args = [ self.name, self._cb_conf['bdriver'], cb, self._cb_cmd_arg, *cb_args ]
+            if cmd and cmd != '-':
+                args = filter(None, args)
```

Comment:
Only `cb_args` holds a variable number of arguments (the comment line above the code is wrong & will be fixed).

So the first four positions are fixed, then follows a variable number. Most script authors should be able to handle that. Otherwise they can stick to the more simple `on_xyz` callbacks that don't receive any arguments.

Reply:
But still - why filter `None`, instead of passing it to the script? It would give the script parameters always at the same positions then.